src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -644,6 +643,60 @@ void FileSegment::completePartAndResetDownloader()
     LOG_TEST(log, "Complete batch. ({})", getInfoForLogUnlocked(lk));
 }
 
+void FileSegment::shrinkFileSegmentToDownloadedSize(const LockedKey & locked_key, const FileSegmentGuard::Lock & lock)
+{
+    chassert(downloaded_size);
+    chassert(fs::file_size(getPath()) > 0);
+
+    if (downloaded_size == range().size())
+    {
+        /// Nothing to resize;
+        return;
+    }
+
+    if (!locked_key.isLastOwnerOfFileSegment(offset()))
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Shrinking of file segment can  be done only by the last holder: {}",
+            getInfoForLog());
+    }
+
+    size_t result_size = downloaded_size;
+
+    const bool shrink_to_aligned_size = !created_from_write_through_cache;
```

Comment:
I'm struggling to understand why we negate `created_from_write_through_cache` here. `CreateFileSegmentSettings seems to be the only place where it will be set to `true`, and hence, we will shrink to `downloaded_size`. But shouldn't it be the opposite?
Btw, can we add a test for this logic pls?

Reply:
it should make code simpler, maybe indeed let's get rid of it?

Reply:
Done